Sorting out issue with calling predict on matrix

pat-alt opened the pull request:
Closes #218

ablaom commented on src/core.jl:
```diff
@@ -144,6 +144,10 @@ nrows(y::AbstractVector) = length(y)
 
+# ---------------------------------
+# Reformatting matrices
+reformat(X, ::Type{<:AbstractMatrix}) = permutedims(X)
```
Yes, when MLJ users provide inputs in the form of a matrix, the observations are rows, which is the opposite of what the internal Flux operations are expecting. 

However, I wouldn't we be better off with and adjoint here, instead of `permute dims`? Then, if a user provides an adjoint, the adjoints cancel and there is no performance cost at all. No?

pat-alt replied:
Yes, good point! Have changed that

ablaom reviewed (commented):
@pat-alt Thanks for looking at this. 

Yes, I think your proposal ought to work: matrices can be provided not only to predict, but also to fit, which we should have for consistency.

What's left to do is:

- [x] Add tests
- [x] Update the scitype declarations
- [x] Update the docstrings

For example, [this code](https://github.com/FluxML/MLJFlux.jl/blob/452c09d7dc13914f4057c661448bc310c9362d3d/src/classifier.jl#L30) should be changed to 

```julia
MLJModelInterface.metadata_model(NeuralNetworkClassifier,
                                 input=Union{AbstractMatrix{Continuous}, Table(Continuous)},
                                 target=AbstractVector{<:Finite},
                                 path="MLJFlux.NeuralNetworkClassifier")
```
Similar changes can be made at regressor.jl

pat-alt commented:
I've updated the scitype declarations for regression ([here](https://github.com/FluxML/MLJFlux.jl/blob/3c369154d85faadd533219bcef79ba4078d7bc54/src/regressor.jl#L23) and [here](https://github.com/FluxML/MLJFlux.jl/blob/3c369154d85faadd533219bcef79ba4078d7bc54/src/regressor.jl#L55)) and for classification ([here](https://github.com/FluxML/MLJFlux.jl/blob/3c369154d85faadd533219bcef79ba4078d7bc54/src/classifier.jl#L30)). For testing, I thought I'd just rerun basic tests with a matrix input (see [here](https://github.com/FluxML/MLJFlux.jl/blob/3c369154d85faadd533219bcef79ba4078d7bc54/test/classifier.jl#L39)). Unfortunately, I can't immediately make sense of the error I'm getting.

ablaom commented:
Thanks for persisting with this.

This error usually means we're trying to grab the `schema` of an object that doesn't have one. I believe the culprit is [here](https://github.com/pat-alt/MLJFlux.jl/blob/3c369154d85faadd533219bcef79ba4078d7bc54/src/classifier.jl#L6). We could add a second method dispatching on matrices to address. There are similar issues for `shape` in regressor.jl.

pat-alt commented:
Thanks! I believe that's fixed now, though I did still get some warnings like this one:

```julia
Warning: The number and/or types of data arguments do not match what the specified model
│ supports. Suppress this type check by specifying `scitype_check_level=0`.
│ 
│ Run `@doc MLJFlux.NeuralNetworkRegressor` to learn more about your model's requirements.
│ 
│ Commonly, but non exclusively, supervised models are constructed using the syntax
│ `machine(model, X, y)` or `machine(model, X, y, w)` while most other models are
│ constructed with `machine(model, X)`.  Here `X` are features, `y` a target, and `w`
│ sample or class weights.
│ 
│ In general, data in `machine(model, data...)` is expected to satisfy
│ 
│     scitype(data) <: MLJ.fit_data_scitype(model)
│ 
│ In the present case:
│ 
│ scitype(data) = Tuple{Table{AbstractVector{Continuous}}, AbstractVector{Continuous}}
│ 
│ fit_data_scitype(model) = Tuple{Union{Table{<:AbstractVector{<:Continuous}}, AbstractMatrix{Continuous}}, AbstractVector{<:Finite}}
```

ablaom commented:
> Thanks! I believe that's fixed now, though I did still get some warnings like this one:

According to the warning you are providing the regressor with a `Continuous` target, rather than a `Finite` one (categorical vector).

ablaom commented:
Sorry, rather the scitype declaration for the regressor is wrong. You need `Continuous` in the `target_scitype` where you appear to have `Finite`.

pat-alt commented:
> Sorry, rather the scitype declaration for the regressor is wrong. You need `Continuous` in the `target_scitype` where you appear to have `Finite`.

You're right, thanks. Fixed that now, but I still get some broken tests (e.g. [here](https://github.com/FluxML/MLJFlux.jl/actions/runs/4674766928/jobs/8279249059?pr=219#step:6:405). Is this expected?

ablaom commented:
> Fixed that now, but I still get some broken tests (e.g. [here](https://github.com/FluxML/MLJFlux.jl/actions/runs/4674766928/jobs/8279249059?pr=219#step:6:405). Is this expected?

Yes. Some of the broken tests are because you are looking at CPU tests, where some tests are excluded and get marked as broken. 

The remaining broken tests are explained here https://github.com/FluxML/MLJFlux.jl/issues/87 .

So no further action on your part needed here. I will finish my review shortly, thanks.

ablaom reviewed (approved):
I'm happy this as is. 

However, perhaps you would consider one additional enhancement, for consistency:
The MulitargetNeuralNetworkRegressor accepts the *target* as a table, but not as a matrix. It seems this would not be difficult to fix (but I haven't checked):

- In [`fitresult(model::MultitargetNeuralNetworkRegressor, chain, y)` method](https://github.com/FluxML/MLJFlux.jl/blob/d01f5366f6cca73d8bd650ac1ab7e4c14735fd77/src/regressor.jl#L53)  return `nothing` instead of `Table.schema(y).names` if `y` is a matrix. 
- In [`shape`](https://github.com/FluxML/MLJFlux.jl/blob/d01f5366f6cca73d8bd650ac1ab7e4c14735fd77/src/regressor.jl#L42) make the change for treating `y` you already have made for `X`.
- In [`predict`](https://github.com/FluxML/MLJFlux.jl/blob/d01f5366f6cca73d8bd650ac1ab7e4c14735fd77/src/regressor.jl#L57) return a matrix if `target_column_names` is `nothing`.  

If you'd rather proceed as we are, then I'll just open an new issue and reference this comment.

pat-alt commented:
Thanks, good idea! I've made the necessary adjustments and updated tests accordingly. Had to make one additional small modification to the `collate` function. I've also updated the docstring.

ablaom commented on src/core.jl:
```diff
@@ -224,6 +224,7 @@ by `model.batch_size`.)
 
 """
 function collate(model, X, y)
+    y = y isa Matrix ? Tables.table(y) : y
```
I'm trying to understand why this fix was needed. The `nrows` function is supposed to work for matrices as well as tables:

```julia
using MLJBase
julia> y = rand(2, 6)
julia> nrows(y)
2

julia> nrows(y')
6
```
And line 230 below should already take care of the conversion of `y` to a matrix, no?

pat-alt replied:
This was giving me an error previously, because the `nrows` method defined in the same file expects a table (or else it throws an `ArgumentError`. I've adjusted that and tests are passing.

ablaom commented on src/regressor.jl:
```diff
@@ -12,23 +18,30 @@ build(model::NeuralNetworkRegressor, rng, shape) =
+
+A private method that returns the shape of the input and output of the model for given data `X` and `y`.
+"""
 function shape(model::MultitargetNeuralNetworkRegressor, X, y)
+    X = X isa Matrix ? Tables.table(X) : X
+    y = y isa Matrix ? Tables.table(y) : y
     n_input = Tables.schema(X).names |> length
     n_output = Tables.schema(y).names |> length
     return (n_input, n_output)
```
Sorry for not catching this earlier, but wrapping a matrix as a table seems a convoluted  way to get num  columns. 

What about something like this?
```suggestion
ncols(X::AbstractMatrix) = size(X, 2)
ncols(X) = Tables.columns(X)) |> Tables.columnnames |> length
shape(model::MultitargetNeuralNetworkRegressor, X, y) = (ncols(X), ncols(y))
```

(Note, I've also dumped `Tables.schema` in favour of different Tables methods, which should be more reliable: `schema` should only be applied to `Tables.rows(X)` or `Tables.columns(X)` and can give `nothing` if not implemented for a particular table. Unfortunately, one cannot avoid `Tables.rows(X)` or `Tables.columns(X)` to get reliably get the number of columns through Tables.jl API, as far as I can understand.)

pat-alt replied:
That makes more sense, yes!

ablaom commented on src/core.jl:
```diff
@@ -141,6 +141,7 @@ function nrows(X)
 end
 nrows(y::AbstractVector) = length(y)
+nrows(X::AbstractMatrix) = size(X, 1)
 
```
Oh, I see. We're not using `nrows` from MLJBase.jl - I forgot. Thanks for humouring me with the explanation!

ablaom reviewed (approved):
Looks good to go.
